pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/PatchUtil.java

```diff
@@ -121,6 +122,7 @@ private static void getColumnNames(Class<?> obj) {
             }
 
             ColumnName[] columnName = field.getAnnotationsByType(ColumnName.class);
+            UniqueField isUniqueField = field.getAnnotation(UniqueField.class);
```

Comment:
I don't understand the intent and scope of `UniqueField`.  Can you say more about what is being checked for uniqueness?  Is this trying to check for uniqueness in an unbounded static map that grows with every unique field DSM tries to insert for however long the VM is running?  This doesn't appear to query the database, so the uniqueness is confined to the VM, and clients have to know to call `getDataBaseMap()` and check it in some way, so it looks like it's easy for client code to ignore this, except for `Patch`, which checks and throws an exception if something isn't unique in the static map.

There is also no synchronization around the static `PatchUtil` fields, which might lead to improper state in these maps under load.

I know this general approach predates the PR, but the lack of tests that show proper functioning of `PatchUtil`'s concepts of uniqueness makes me wonder if this is going to solve the issue, or if it's better to leave the indexes in place and look at `equals()` and `hashCode()` to help client code remove duplicates before persisting to the db.

Should we be thinking about an upsert approach in `Patch` that takes into account the `deleted` column?

At the end of the day, if something is marked as `unique` by all of this, the external behavior change is the throwing of `DuplicateException`, which we had removed from the codebase recently because nothing responds to it in production--it's an exception without any different handling (although I do see that a new test does check for it).

Reply:
Of course, let me explain better: 
 > What is the specific purpose of UniqueField within our codebase? Can you provide a more detailed explanation of its intended functionality?

The primary purpose of the `UniqueField` within our codebase is to ensure the uniqueness of fields while exempting records marked as 'deleted'. The motivation behind this feature is to enable reuse of data from deleted records without triggering SQL duplicate exceptions. To illustrate, consider a scenario where a CRC erroneously deletes a tissue entry. With this implementation, they should be able to recreate the tissue, reusing the same data, without any hitches. Consequently, to accommodate this functionality, our current unique indexes in the tables would need to be removed. The UniqueField annotation informs the Patch Service about fields that should maintain uniqueness across all undeleted data. Before any insertion, the system verifies whether the data already exists, acting as a "soft check" on the field's uniqueness.

Reply:
> Is this trying to check for uniqueness in an unbounded static map that grows with every unique field DSM tries to insert for however long the VM is running?

UniqueField employs a static map initialized during server startup. This map plays a crucial role in all patch requests, serving primarily to associate frontend field names with their corresponding backend columns and tables. The incorporation of the UniqueField annotation merely adds an extra characteristic to this existing mapping process.

The uniqueness check operates through this static map. If an entry corresponds to a field marked with this annotation, the system prompts Patch to verify the value's presence among all "not deleted" records—effectively excluding any records marked with deleted=true—before proceeding with an insertion.

It's important to note that this approach shouldn't impose additional load on the server. The static map is a foundational component that’s already employed for other operations, and the UniqueField simply integrates an extra layer of validation within this existing framework. Consequently, while we're enhancing data integrity by ensuring uniqueness where required, the overhead is negligible because we're capitalizing on pre-existing infrastructure (the static map) that's fundamental to the system’s operation.